### PR TITLE
accuraterip-checksum: convert to python C extension

### PR DESCRIPTION
**Pull request by mtdcr**
* calculate v1 and v2 checksums at once
* let libsndfile handle both WAV and FLAC
---
I tested this using the following script:
```
#!/usr/bin/env python

import accuraterip

v1, v2 = accuraterip.compute('Nirvana - Nevermind (Remastered)/01. Nirvana - Smells Like Teen Spirit.flac', 1, 12)
assert(v1 == 0xA0D81775)
assert(v2 == 0x280694C8)

v1, v2 = accuraterip.compute('Nirvana - Nevermind (Remastered)/02. Nirvana - In Bloom.flac', 2, 12)
assert(v1 == 0x92EC1D7D)
assert(v2 == 0x1AE5FD7A)

v1, v2 = accuraterip.compute('Nirvana - Nevermind (Remastered)/12. Nirvana - Something in the Way - Endless, Nameless.flac', 12, 12)
assert(v1 == 0x3EAA3B67)
assert(v2 == 0x14C921A5)

print("OK")
```
This may need some more testing. When called with invalid input files, `compute()` returns `None, None`. Note that 0 may be a valid checksum, so the statement `if not v1_sum` in accurip.py may have produced false negatives in the past.

### CODE REVIEW

**Comment by mtdcr** [comment]
I updated my branch to fix the encoding of the filename passed to accuraterip.compute().

**Comment by RecursiveForest** [comment]
I like this. I will give this a more thorough review & testing shortly. A simple to answer question is whether the existing unit tests are testing this properly or not.

**Comment by MerlijnWajer** [comment]
What is the advantage over a binary - unix style? Seems more complex to me.

**Comment by mtdcr** [comment]
My commit removes more lines than it adds or modifies. Therefore I'd argue that complexity actually gets reduced.

Benefits are:
* Both checksums (v1 and v2) are calculated at once, reducing execution time and use of resources.
* No need to run make to install the executable, because the module gets installed by setup.py just like the rest of whipper.
* Simple and stable interface between native Python code and the module.
* Supports FLAC directly through libsndfile, removing the conversion step from flac to wav before calling the executable.

Please have a look at how simple arc.py became!

It would be possible to simplify accurip.py a little more, because the module returns either both checksums or none, whereas the executable could fail to calculate one checksum but succeed with the other (if the file becomes unreadable between both invocations), but I preferred to keep the commit small.

**Comment by MerlijnWajer** [comment]
I would say that (1) can be done in the binary quite easily too - but I never did it because the speedup was so significant I didn't need to make it more complex. (2) is fair, but binaries can also be installed by setup.py AFAIK (3) I would say that simple program output + exit code is also a stable interface (4) I purposefully re-used the flac binary, so that was a feature. ;) 

I personally find python modules cumbersome to debug (more so than simply binaries for sure), but I don' feel that strongly about this in general and the interface looks slightly more simple (but a simple interface can hide a lot of complexity), so if people want this merged, it's fine by me.

**Comment by MerlijnWajer** [comment]
Oh, and thanks for taking the time to write this.

**Comment by mtdcr** [comment]
@RecursiveForest, are you going to review this PR?

**Comment by mtdcr** [comment]
Regarding your question about the unit tests: I guess they do cover it. See test_common_accurip.py.

File: src/accuraterip-checksum.c [hunk 1]
```diff
@@ -2,11 +2,11 @@
  ============================================================================
  Name        : accuraterip-checksum.c
  Author      : Leo Bogert (http://leo.bogert.de)
- Git         : http://leo.bogert.de/accuraterip-checksum
- Version     : See global variable "version"
- Copyright   : GPL
- Description : A C99 commandline program to compute the AccurateRip checksum of singletrack WAV files.
- 	 	 	   Implemented according to http://www.hydrogenaudio.org/forums/index.php?showtopic=97603
+ Git         : https://github.com/leo-bogert/accuraterip-checksum
```
Comment (Freso) [inline src/accuraterip-checksum.c]: This is no longer accurate: the code in this `accuraterip-checksum.c` is not the same in the linked repository.

**Comment by mtdcr** [comment]
> This PR introduced a glob of C code unique to this project, which means any maintenance would have to happen in the scope of whipper as well.

I don't get your point. The "original" author didn't seem to understand C very well either. The current checksum routines were verbatim copies from a message board. Therefore there's no external maintainer and thus maintenance already happens in the scope of whipper now. Well, actually maintenance didn't happen anywhere at all, until this PR.

Second, IIRC, there was an implementation in Python, either in whipper or its predecessor, which was replaced with C code due to its bad performance. See MerlijnWajer's second comment in this PR.

Of course I'm not arguing to set this module in stone and keep it forever. Keeping it local to whipper's repository gives you the freedom to replace it with any superior solution you're going to invent in the future. And I don't even think it would be hard to do, but it may certainly be done in a second iteration.

This PR just improves the status quo a little bit, because I actually took the time to try and understand how the checksum works (and how the old checksum differs from the newer one), which the original author apparently didn't. You can even use this new code as a blueprint for efficient Python code later.

Btw., I rebased the patch locally the last time I ripped a CD. It's been a while, but I can push an updated version, if anybody is interested.

And, frankly, I don't understand why moving it out of the whipper repository would improve anything. It would just live under https://github.com/whipper-team/<obscure-name> and have the same maintainers as now.

**Comment by spvkgn** [comment]
> Btw., I rebased the patch locally the last time I ripped a CD. It's been a while, but I can push an updated version, if anybody is interested.

I'm use this PR patch for whipper deb package on Launchpad [PPA](https://launchpad.net/~spvkgn/+archive/ubuntu/whipper). It would be useful if you update your branch with 0.7.3 sources. Thanks!

**Comment by MerlijnWajer** [comment]
Would you be up for creating an executable and also a python extension, sharing the same core?

I definitely like code and performance improvements. And yeah, the current code (in whipper) is not stellar, I didn't bother cleaning up the code, I just fixed it to work for my use case.

I like the idea of cleaned up code, I just prefer to have the feature available as a binary (as well, at least).

**Comment by mtdcr** [comment]
@spvkgn: Sorry, I was busy with other things and forgot about this PR. Few minutes ago, I rebased it on today's develop branch. This version didn't receive any testing, but there were no significant changes since the previous attempt.

@Freso: I removed the reference to the former Git repository as requested.

@MerlijnWajer: I don't think an additional executable would improve anything, but confuse new users. Couldn't you just use https://github.com/leo-bogert/accuraterip-checksum directly, if you needed it outside of whipper?

**Comment by MerlijnWajer** [comment]
How would it confuse new users? Using the repo you linked directly at this point differs significantly from the code in this PR.

**Comment by mtdcr** [comment]
@MerlijnWajer How about this script as a replacement for your binary? Is there anything else blocking this PR?

``` python
#!/usr/bin/env python
# SPDX-License-Identifier: GPL-3.0-only

import accuraterip
import sys


if len(sys.argv) == 2 and sys.argv[1] == '--version':
    print('accuraterip-checksum version 2.0')
    exit(0)

use_v1 = None
if len(sys.argv) == 4:
    offset = 0
    use_v1 = False
elif len(sys.argv) == 5:
    offset = 1
    if sys.argv[1] == '--accuraterip-v1':
        use_v1 = True
    elif sys.argv[1] == '--accuraterip-v2':
        use_v1 = False

if use_v1 is None:
    print('Syntax: accuraterip-checksum [--version / --accuraterip-v1 / --accuraterip-v2 (default)] filename track_number total_tracks')
    exit(1)

filename = sys.argv[offset + 1]
track_number = int(sys.argv[offset + 2])
total_tracks = int(sys.argv[offset + 3])

v1, v2 = accuraterip.compute(filename, track_number, total_tracks)
if use_v1:
    print('%08X' % v1)
else:
    print('%08X' % v2)
```

**Comment by MerlijnWajer** [comment]
That looks clean. Can you add that to setup.py so it gets installed as executable?

**Comment by mtdcr** [comment]
@MerlijnWajer: Done.

@RecursiveForest: Could you please review this PR or remove yourself from the list of pending reviewers?

**Comment by JoeLametta** [comment]
/remind me to remove RecursiveForest from the list of reviewers if we get no reply in 1 week.

**Comment by JoeLametta** [comment]
/remind me to remove RecursiveForest from the list of reviewers if we get no reply in 1 week

**Comment by MerlijnWajer** [comment]
@JoeLametta - I'm good with getting this merged whenever.

**Comment by MerlijnWajer** [comment]
Thanks for the work and putting up with me, @mtdcr ! :)

**Comment by JoeLametta** [comment]
Merged, thanks!

**Comment by ArchangeGabriel** [comment]
So just to be sure, this means we should not use an external accuraterip-checksum anymore, right?

**Comment by JoeLametta** [comment]
@mtdcr Hi, starting with the next release whipper will be _Python 3 **only**_ (porting progress is in #411). It seems that the current code of the `accuraterip-checksum` extension is not compatible with Python 3:

```
import accuraterip

ImportError: /home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/whipper-0.7.4.dev109+gd626c9c-py3.5-linux-x86_64.egg/accuraterip.cpython-35m-x86_64-linux-gnu.so: undefined symbol: Py_InitModule
```

Could you help us porting it to Python 3?

Thanks,
Joe

**Comment by JoeLametta** [comment]
> So just to be sure, this means we should not use an external accuraterip-checksum anymore, right?

Correct.

**Comment by mtdcr** [comment]
> `undefined symbol: Py_InitModule`

@JoeLametta can you try this patch? It passes `python3 setup.py build` now, but I didn't try to execute it.
https://github.com/mtdcr/whipper/commit/5183f93752ea7e68045b48deaad277318c1a0e34

**Comment by JoeLametta** [comment]
I've merged the patch in the `py3` branch (commit https://github.com/whipper-team/whipper/pull/411/commits/0fc330ce7e58300f2ddd4a9ad07060bf8eeb6206): according to [Travis CI](https://travis-ci.com/whipper-team/whipper/jobs/251207645) the tests related to `accuraterip-checksum` aren't failing anymore. :+1:

**Comment by ArchangeGabriel** [comment]
@JoeLametta Thanks. Also I think that [this part of the instructions](https://github.com/whipper-team/whipper#building-the-bundled-dependencies) should be removed since the setup.py already handles this (and the Makefile is gone for instance).

**Comment by JoeLametta** [comment]
> I think that [this part of the instructions](https://github.com/whipper-team/whipper#building-the-bundled-dependencies) should be removed since the setup.py already handles this (and the Makefile is gone for instance).

Done! :ok_hand:

**Comment by mtdcr** [comment]
> > I think that [this part of the instructions](https://github.com/whipper-team/whipper#building-the-bundled-dependencies) should be removed since the setup.py already handles this (and the Makefile is gone for instance).
> 
> Done!

Did you forget to push it to develop?

**Comment by ArchangeGabriel** [comment]
No, he pushed as part of the Python3 porting PR #411.

**Comment by JoeLametta** [comment]
> No, he pushed as part of the Python3 porting PR #411.

That's right (https://github.com/whipper-team/whipper/pull/411/commits/9edc98ddd0ec33e2945b8706bf3bc4d1c9f136a3).

**Pull request merged**